pkg/dataobj/querier/iter.go

```diff
@@ -25,8 +28,253 @@ var (
 			return &samples
 		},
 	}
+	entryWithLabelsPool = sync.Pool{
+		New: func() interface{} {
+			entries := make([]entryWithLabels, 0, 1024)
+			return &entries
+		},
+	}
 )
 
+type entryWithLabels struct {
+	Labels     string
+	StreamHash uint64
+	Entry      logproto.Entry
+}
+
+// newEntryIterator creates a new EntryIterator for the given context, streams, and reader.
+// It reads records from the reader and adds them to the topk heap based on the direction.
+// The topk heap is used to maintain the top k entries based on the direction.
+// The final result is returned as a slice of entries.
+func newEntryIterator(ctx context.Context,
+	streams map[int64]dataobj.Stream,
+	reader *dataobj.LogsReader,
+	req logql.SelectLogParams,
+) (iter.EntryIterator, error) {
+	bufPtr := recordsPool.Get().(*[]dataobj.Record)
+	defer recordsPool.Put(bufPtr)
+	buf := *bufPtr
+
+	selector, err := req.LogSelector()
+	if err != nil {
+		return nil, err
+	}
+	pipeline, err := selector.Pipeline()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		prevStreamID    int64 = -1
+		streamExtractor log.StreamPipeline
+		streamHash      uint64
+		top             = newTopK(int(req.Limit), req.Direction)
+	)
+
+	for {
+		n, err := reader.Read(ctx, buf)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+
+		if n == 0 {
```

Comment:
You'll need to only break if n == 0 and you got EOF. Especially with the changes coming in soon, there may be some rare calls where n == 0 but there's still more data to read (ie if all data got filtered out in a batch)

Reply:
ok although you should keep going if the batch is empty ?

Reply:
sending 0 without EOF I don't think it's similar to io.Reader

Reply:
Actually seems like totally normal !